client/rpmtrans.c

```diff
@@ -211,13 +211,115 @@ TDNFPopulateTransaction(
     goto cleanup;
 }
 
+uint32_t
+TDNFDetectPreTransFailure(
+    rpmts pTS,
+    char *pszError
+    )
+{
+    uint32_t dwError = 0;
+    rpmtsi pi = NULL;
+    rpmte pte = NULL;
+    char *pszToken;
+    int i = 0;
+    char *pszPkgName = NULL;
+    char *pszSymbol = NULL;
+    char *pszVersion = NULL;
+    char *pszEVR = NULL;
+    char *pszCachePkgName = NULL;
+    char *pszCachePkgEVR = NULL;
+
+    if (!pTS || !pszError)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+    // Error Str has the format: <Pkg-Name> <Symbol> <version-release>
+    pszToken = strtok(pszError, " ");
+    while (pszToken != NULL)
+    {
+        switch(i)
+        {
+            case 0:
+                pszPkgName = pszToken;
+                break;
+            case 1:
+                pszSymbol = pszToken;
+                break;
+            case 2:
+                pszVersion = pszToken;
+                break;
+        }
+        i++;
+        pszToken = strtok(NULL, " ");
+    }
+
+    if (!pszPkgName || !pszSymbol || !pszVersion)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    pi = rpmtsiInit(pTS);
+    while ((pte = rpmtsiNext(pi, 0)) != NULL)
+    {
+        dwError = TDNFAllocateString(rpmteN(pte), &pszCachePkgName);
+        BAIL_ON_TDNF_ERROR(dwError);
+
+        dwError = TDNFAllocateString(rpmteEVR(pte), &pszCachePkgEVR);
+        BAIL_ON_TDNF_ERROR(dwError);
+
+        pszEVR = strtok(pszCachePkgEVR, "ph");
+        if (!pszEVR)
+        {
+            dwError = ERROR_TDNF_INVALID_PARAMETER;
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+
+        if (strcmp(pszCachePkgName, pszPkgName) == 0)
+        {
+            if (strstr(pszSymbol, ">") && (rpmvercmp(pszEVR, pszVersion) > 0))
+            {
+                fprintf(stderr, "Detected rpm pre-transaction dependency errors."
+                        "Install %s %s %s separately first to resolve this failure.\n",
+                        pszPkgName, pszSymbol, pszVersion);
+                break;
+            }
+            else if (strstr(pszSymbol, "<") && (rpmvercmp(pszEVR, pszVersion) < 0))
+            {
+                fprintf(stderr, "Detected rpm pre-transaction dependency errors."
+                        "Install %s %s %s separately first to resolve this failure.\n",
+                        pszPkgName, pszSymbol, pszVersion);
+                break;
+            }
+            else if (strstr(pszSymbol, "=") && (rpmvercmp(pszEVR, pszVersion) == 0))
+            {
+                fprintf(stderr, "Detected rpm pre-transaction dependency errors."
+                        "Install %s %s %s separately first to resolve this failure.\n",
+                        pszPkgName, pszSymbol, pszVersion);
+                break;
+            }
+        }
+    }
+    rpmtsiFree(pi);
```

Comment:
Does this need to be freed in case of error? I mean, it should be added in cleanup:?

Reply:
will do that. Thanks